lib/pure/options.nim

```diff
@@ -379,3 +379,122 @@ proc unsafeGet*[T](self: Option[T]): lent T {.inline.}=
   ## Generally, using the `get proc <#get,Option[T]>`_ is preferred.
   assert self.isSome
   result = self.val
+
+template mapIt*[T](value: Option[T], action: untyped): untyped =
+  ## Applies an action to the value of the `Option`, if it has one.
+  runnableExamples:
+    assert some(42).mapIt(it * 2).mapIt($it) == some("84")
+    assert none(int).mapIt(it * 2).mapIt($it) == none(string)
+
+  block:
+    type InnerType = typeof(
+      block:
+        var it {.inject, used.}: typeof(value.get())
+        action
+    )
+
+    var outcome: Option[InnerType]
+    if value.isSome:
+      let it {.inject, used.} = unsafeGet(value)
+      outcome = some(action)
+    outcome
+
+template flatMapIt*[T](value: Option[T], action: untyped): untyped =
+  ## Executes an action on the value of the `Option`, where that action can also return an `Option`.
+  runnableExamples:
+    assert some(42).flatMapIt(some($it)) == some("42")
+    assert some(42).flatMapIt(none(string)) == none(string)
+    assert none(int).flatMapIt(some($it)) == none(string)
+    assert none(int).flatMapIt(none(string)) == none(string)
+
+  block:
+    type InnerType = typeof(
+      block:
+        var it {.inject, used.}: typeof(value.get())
+        action.get()
+    )
+
+    var outcome: Option[InnerType]
+    if value.isSome:
+      let it {.inject, used.} = value.get()
+      outcome = action
+    outcome
+
+template filterIt*[T](value: Option[T], action: untyped): Option[T] =
+  ## Tests the value of the `Option` with a predicate, returning a `none` if it fails.
+  runnableExamples:
+    assert some(42).filterIt(it > 0) == some(42)
+    assert none(int).filterIt(it > 0) == none(int)
+    assert some(-11).filterIt(it > 0) == none(int)
+  block:
+    var outcome: Option[T]
+    if value.isSome:
+      let it {.inject, used.} = value.unsafeGet
+      if action:
+        outcome = value
+    outcome
+
+template applyIt*[T](value: Option[T], action: untyped) =
+  ## Executes a code block if the `Option` is `some`, assigning the value to a variable named `it`
+  runnableExamples:
+    var value: string
+    some("foo").applyIt:
+      value = it
+    assert value == "foo"
+
+    none(string).applyIt:
+      assert false
+
+  block:
+    if value.isSome:
+      let it {.inject, used.} = value.unsafeGet
+      action
+
+template orElse*[T](value: Option[T], otherwise: untyped): auto =
+  ## Returns the value in an option if it is set. Otherwise, executes a code block. This is
+  ## useful for executing side effects when the option is empty.
+  runnableExamples:
+    let value = some("foo").orElse:
+      assert false
+    assert value == "foo"
+
+    block foo:
+      discard none(string).orElse:
+        break foo
+      assert false
+
+  block:
+    var outcome: T
+    if value.isSome:
+      outcome = value.unsafeGet()
+    else:
+      otherwise
+    outcome
+
+template `from`*[T](variable: untyped, source: Option[T]): bool =
```

Comment:
Too alien. Needs a better name/syntax.

Reply:
:+1:  I'll probably go with the `withValue` pattern used by the `tables` stdlib